File: frontend/lib/api/domains/projects.ts
/**
 * Projects Domain API
 * Includes: Projects, Tasks, Dashboard, Analytics
 */

import { fetchApi } from '../core';

// =============================================================================
// TYPES
// =============================================================================

export type ProjectStatus = 'open' | 'completed' | 'cancelled' | 'on_hold';
export type ProjectPriority = 'low' | 'medium' | 'high';

export interface ProjectUser {
  user?: string | null;
  full_name?: string | null;
  email?: string | null;
  project_status?: string | null;
  view_attachments?: boolean;
  welcome_email_sent?: boolean;
  idx?: number;
}

export interface ProjectListItem {
  id: number;
  erpnext_id: string | null;
  project_name: string;
  project_type: string | null;
  status: ProjectStatus;
  priority: ProjectPriority | null;
  department: string | null;
  customer_id: number | null;
  percent_complete: number | null;
  expected_start_date: string | null;
  expected_end_date: string | null;
  estimated_costing: number | null;
  total_billed_amount: number | null;
  is_overdue: boolean;
  task_count: number | null;
  created_at: string | null;
  write_back_status?: string | null;
}

export interface ProjectListResponse {
  total: number;
  limit: number;
  offset: number;
  data: ProjectListItem[];
}

export interface ProjectDetail extends ProjectListItem {
  company?: string | null;
  cost_center?: string | null;
  project_manager_id?: number | null;
  project_manager?: string | null;
  erpnext_customer?: string | null;
  erpnext_sales_order?: string | null;
  percent_complete_method?: string | null;
  is_active?: string | null;
  actual_time?: number | null;
  total_consumed_material_cost?: number | null;
  total_costing_amount?: number | null;
  total_expense_claim?: number | null;
  total_purchase_cost?: number | null;
  total_sales_amount?: number | null;
  total_billable_amount?: number | null;
  gross_margin?: number | null;
  per_gross_margin?: number | null;
  collect_progress?: boolean;
  frequency?: string | null;
  message?: string | null;
  notes?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
  from_time?: string | null;
  to_time?: string | null;
  customer?: Record<string, unknown> | null;
  users?: ProjectUser[];
  tasks?: unknown[];
  task_stats?: Record<string, unknown>;
  expenses?: unknown[];
  time_tracking?: Record<string, unknown>;
}

export interface ProjectPayload {
  project_name?: string;
  project_type?: string | null;
  status?: ProjectStatus;
  priority?: ProjectPriority;
  department?: string | null;
  company?: string | null;
  cost_center?: string | null;
  customer_id?: number | null;
  project_manager_id?: number | null;
  erpnext_customer?: string | null;
  erpnext_sales_order?: string | null;
  percent_complete?: number | null;
  percent_complete_method?: string | null;
  is_active?: string | null;
  actual_time?: number | null;
  total_consumed_material_cost?: number | null;
  estimated_costing?: number | null;
  total_costing_amount?: number | null;
  total_expense_claim?: number | null;
  total_purchase_cost?: number | null;
  total_sales_amount?: number | null;
  total_billable_amount?: number | null;
  total_billed_amount?: number | null;
  gross_margin?: number | null;
  per_gross_margin?: number | null;
  collect_progress?: boolean;
  frequency?: string | null;
  message?: string | null;
  notes?: string | null;
  expected_start_date?: string | null;
  expected_end_date?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
  from_time?: string | null;
  to_time?: string | null;
  users?: ProjectUser[];
}

export interface ProjectsDashboard {
  cards?: Record<string, number>;
  projects?: {
    total?: number;
    active?: number;
    completed?: number;
    on_hold?: number;
    cancelled?: number;
  };
  tasks?: {
    total?: number;
    open?: number;
    overdue?: number;
  };
  financials?: {
    total_billed?: number;
  };
  metrics?: {
    avg_completion_percent?: number;
    due_this_week?: number;
    [key: string]: unknown;
  };
  by_priority?: Record<string, number>;
}

export interface ProjectTaskListResponse {
  total: number;
  limit: number;
  offset: number;
  data: unknown[];
}

export interface ProjectListParams {
  status?: string;
  priority?: ProjectPriority;
  customer_id?: number;
  project_type?: string;
  department?: string;
  search?: string;
  overdue_only?: boolean;
  start_date?: string;
  end_date?: string;
  limit?: number;
  offset?: number;
}

export interface ProjectTaskParams {
  project_id?: number;
  status?: string;
  priority?: string;
  assigned_to?: string;
  task_type?: string;
  search?: string;
  overdue_only?: boolean;
  start_date?: string;
  end_date?: string;
  limit?: number;
  offset?: number;
}

export interface ProjectTaskUpdatePayload {
  assigned_to?: string;
}

export interface ProjectTaskCreatePayload {
  subject: string;
  project_id?: number;
  status?: string;
  priority?: string;
  assigned_to?: string;
  task_type?: string;
  exp_start_date?: string;
  exp_end_date?: string;
  description?: string;
}

export interface ProjectTask {
  id: number;
  subject?: string;
}

// =============================================================================
// MILESTONE TYPES
// =============================================================================

export type MilestoneStatus = 'planned' | 'in_progress' | 'completed' | 'on_hold';

export interface Milestone {
  id: number;
  project_id: number;
  name: string;
  description?: string | null;
  status: MilestoneStatus;
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
  percent_complete: number;
  idx: number;
  is_overdue: boolean;
  task_count: number;
  tasks?: Array<{
    id: number;
    erpnext_id?: string | null;
    subject: string;
    status?: string | null;
    priority?: string | null;
    assigned_to?: string | null;
    progress: number;
    exp_start_date?: string | null;
    exp_end_date?: string | null;
    is_overdue: boolean;
  }>;
  created_by_id?: number | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface MilestoneListResponse {
  total: number;
  data: Milestone[];
}

export interface MilestoneCreatePayload {
  name: string;
  description?: string | null;
  status?: MilestoneStatus;
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
  percent_complete?: number;
  idx?: number;
}

export interface MilestoneUpdatePayload {
  name?: string;
  description?: string | null;
  status?: MilestoneStatus;
  planned_start_date?: string | null;
  planned_end_date?: string | null;
  actual_start_date?: string | null;
  actual_end_date?: string | null;
  percent_complete?: number;
  idx?: number;
}

// =============================================================================
// COMMENT TYPES
// =============================================================================

export type EntityType = 'project' | 'task' | 'milestone';

export interface ProjectComment {
  id: number;
  entity_type: EntityType;
  entity_id: number;
  content: string;
  author_id: number;
  author_name?: string | null;
  author_email?: string | null;
  is_edited: boolean;
  edited_at?: string | null;
  created_at?: string | null;
}

export interface CommentListResponse {
  total: number;
  limit: number;
  offset: number;
  data: ProjectComment[];
}

export interface CommentCreatePayload {
  content: string;
}

export interface CommentUpdatePayload {
  content: string;
}

// =============================================================================
// ACTIVITY TYPES
// =============================================================================

export type ProjectActivityType =
  | 'created'
  | 'updated'
  | 'status_changed'
  | 'assigned'
  | 'comment_added'
  | 'attachment_added'
  | 'milestone_completed'
  | 'task_completed'
  | 'approval_submitted'
  | 'approval_approved'
  | 'approval_rejected';

export interface ProjectActivity {
  id: number;
  entity_type: EntityType;
  entity_id: number;
  activity_type: ProjectActivityType;
  description: string;
  from_value?: string | null;
  to_value?: string | null;
  changed_fields?: string[] | null;
  actor_id?: number | null;
  actor_name?: string | null;
  actor_email?: string | null;
  created_at?: string | null;
}

export interface ActivityListResponse {
  total: number;
  limit: number;
  offset: number;
  data: ProjectActivity[];
}

export interface ActivityTimelineResponse {
  project_id: number;
  total: number;
  data: ProjectActivity[];
}

// =============================================================================
// ATTACHMENT TYPES
// =============================================================================

export interface ProjectAttachment {
  id: number;
  doctype?: string;
  document_id?: number;
  file_name: string;
  file_path?: string;
  file_type?: string | null;
  file_size?: number | null;
  attachment_type?: string | null;
  is_primary: boolean;
  description?: string | null;
  uploaded_at?: string | null;
  uploaded_by_id?: number | null;
}

export interface AttachmentListResponse {
  total: number;
  data: ProjectAttachment[];
}

// =============================================================================
// APPROVAL TYPES
// =============================================================================

export type ApprovalStatusType = 'draft' | 'pending' | 'approved' | 'rejected' | 'cancelled' | 'posted';

// =============================================================================
// TEMPLATE TYPES
// =============================================================================

export interface MilestoneTemplateItem {
  id: number;
  project_template_id: number;
  name: string;
  description?: string | null;
  start_day_offset: number;
  end_day_offset: number;
  idx: number;
}

export interface TaskTemplateItem {
  id: number;
  project_template_id: number;
  subject: string;
  description?: string | null;
  priority: ProjectPriority;
  start_day_offset: number;
  duration_days: number;
  default_assigned_role?: string | null;
  parent_template_id?: number | null;
  milestone_template_id?: number | null;
  is_group: boolean;
  idx: number;
}

export interface ProjectTemplate {
  id: number;
  name: string;
  description?: string | null;
  project_type?: string | null;
  default_priority: ProjectPriority;
  estimated_duration_days?: number | null;
  default_notes?: string | null;
  is_active: boolean;
  company?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
  milestone_templates?: MilestoneTemplateItem[];
  task_templates?: TaskTemplateItem[];
  milestones?: MilestoneTemplateItem[];
  tasks?: TaskTemplateItem[];
  milestone_count?: number;
  task_count?: number;
}

export interface ProjectTemplateListResponse {
  total: number;
  data: ProjectTemplate[];
}

export interface MilestoneTemplatePayload {
  name: string;
  description?: string | null;
  start_day_offset?: number;
  end_day_offset?: number;
  idx?: number;
}

export interface TaskTemplatePayload {
  subject: string;
  description?: string | null;
  priority?: ProjectPriority;
  start_day_offset?: number;
  duration_days?: number;
  default_assigned_role?: string | null;
  parent_template_id?: number | null;
  milestone_template_id?: number | null;
  is_group?: boolean;
  idx?: number;
}

export interface ProjectTemplateCreatePayload {
  name: string;
  description?: string | null;
  project_type?: string | null;
  default_priority?: ProjectPriority;
  estimated_duration_days?: number | null;
  default_notes?: string | null;
  is_active?: boolean;
  milestone_templates?: MilestoneTemplatePayload[];
  task_templates?: TaskTemplatePayload[];
}

export interface ProjectTemplateUpdatePayload {
  name?: string;
  description?: string | null;
  project_type?: string | null;
  default_priority?: ProjectPriority;
  estimated_duration_days?: number | null;
  default_notes?: string | null;
  is_active?: boolean;
  milestone_templates?: MilestoneTemplatePayload[];
  task_templates?: TaskTemplatePayload[];
}

export interface CreateFromTemplatePayload {
  project_name: string;
  expected_start_date?: string;
  customer_id?: number | null;
  project_manager_id?: number | null;
  notes?: string | null;
}

export interface ApprovalHistoryItem {
  step_order: number;
  action: string;
  user_id: number;
  remarks?: string | null;
  action_at: string;
}

export interface ProjectApprovalStatus {
  project_id: number;
  has_approval: boolean;
  approval_id?: number;
  status?: ApprovalStatusType | null;
  current_step?: number | null;
  current_step_name?: string | null;
  amount?: string | null;
  submitted_at?: string | null;
  submitted_by_id?: number | null;
  approved_at?: string | null;
  approved_by_id?: number | null;
  rejected_at?: string | null;
  rejected_by_id?: number | null;
  rejection_reason?: string | null;
  posted_at?: string | null;
  posted_by_id?: number | null;
  message?: string;
  history?: ApprovalHistoryItem[];
}

// =============================================================================
// CHANGE HISTORY TYPES
// =============================================================================

export type HistorySource = 'audit' | 'activity';

export interface ChangeHistoryItem {
  id: string;
  source: HistorySource;
  timestamp?: string | null;
  action?: string | null;
  actor_id?: number | null;
  actor_name?: string | null;
  actor_email?: string | null;
  description?: string | null;
  changed_fields?: string[] | null;
  old_values?: Record<string, unknown> | null;
  new_values?: Record<string, unknown> | null;
  from_value?: string | null;
  to_value?: string | null;
  remarks?: string | null;
}

export interface ChangeHistoryResponse {
  entity_type: EntityType;
  entity_id: number;
  total: number;
  data: ChangeHistoryItem[];
}

// =============================================================================
// API
// =============================================================================

export const projectsApi = {
  // =========================================================================
  // PROJECTS
  // =========================================================================

  getProjects: (params?: ProjectListParams) =>
    fetchApi<ProjectListResponse>('/projects', {
      params: params ? ({ ...params } as Record<string, unknown>) : undefined,
    }),

  getProjectDetail: (id: number) => fetchApi<ProjectDetail>(`/projects/${id}`),

  createProject: (body: ProjectPayload & { project_name: string }) =>
    fetchApi<ProjectDetail>('/projects', { method: 'POST', body: JSON.stringify(body) }),

  updateProject: (id: number, body: ProjectPayload) =>
    fetchApi<ProjectDetail>(`/projects/${id}`, { method: 'PATCH', body: JSON.stringify(body) }),

  deleteProject: (id: number) =>
    fetchApi<{ message: string; id: number }>(`/projects/${id}`, { method: 'DELETE' }),

  // =========================================================================
  // DASHBOARD & ANALYTICS
  // =========================================================================

  getProjectsDashboard: () => fetchApi<ProjectsDashboard>('/projects/dashboard'),

  getProjectsStatusTrend: (months = 12) =>
    fetchApi<unknown>('/projects/analytics/status-trend', { params: { months } }),

  getProjectsTaskDistribution: () =>
    fetchApi<unknown>('/projects/analytics/task-distribution'),

  getProjectsPerformance: () => fetchApi<unknown>('/projects/analytics/project-performance'),

  getProjectsDepartmentSummary: (months = 12) =>
    fetchApi<unknown>('/projects/analytics/department-summary', { params: { months } }),

  // =========================================================================
  // TASKS
  // =========================================================================

  getProjectTasks: (params?: ProjectTaskParams) =>
    fetchApi<ProjectTaskListResponse>('/projects/tasks', {
      params: params ? ({ ...params } as Record<string, unknown>) : undefined,
    }),

  createTask: (body: ProjectTaskCreatePayload) =>
    fetchApi<ProjectTask>('/projects/tasks', {
      method: 'POST',
      body: JSON.stringify(body),
    }),

  getTaskDetail: (id: number) => fetchApi<unknown>(`/projects/tasks/${id}`),

  updateTask: (id: number, body: ProjectTaskUpdatePayload) =>
    fetchApi<unknown>(`/projects/tasks/${id}`, {
      method: 'PATCH',
      body: JSON.stringify(body),
    }),

  // =========================================================================
  // GANTT CHART
  // =========================================================================

  /**
   * Get tasks with full dependency data for Gantt chart.
   * Returns all tasks for a project with their dependencies.
   */
  getProjectGanttData: (projectId: number) =>
    fetchApi<{
      tasks: Array<{
        id: number;
        subject: string;
        status: string;
        priority: string;
        progress: number | null;
        exp_start_date: string | null;
        exp_end_date: string | null;
        assigned_to: string | null;
        parent_task_id: number | null;
        is_group: boolean;
        depends_on: number[];
      }>;
      date_range: {
        min_date: string | null;
        max_date: string | null;
      };
    }>(`/projects/${projectId}/gantt`),

  // =========================================================================
  // MILESTONES
  // =========================================================================

  /** List all milestones for a project */
  getProjectMilestones: (projectId: number, status?: MilestoneStatus) =>
    fetchApi<MilestoneListResponse>(`/projects/${projectId}/milestones`, {
      params: status ? { status } : undefined,
    }),

  /** Get a specific milestone with its tasks */
  getMilestone: (milestoneId: number) =>
    fetchApi<Milestone>(`/projects/milestones/${milestoneId}`),

  /** Create a new milestone */
  createMilestone: (projectId: number, body: MilestoneCreatePayload) =>
    fetchApi<Milestone>(`/projects/${projectId}/milestones`, {
      method: 'POST',
      body: JSON.stringify(body),
    }),

  /** Update a milestone */
  updateMilestone: (milestoneId: number, body: MilestoneUpdatePayload) =>
    fetchApi<Milestone>(`/projects/milestones/${milestoneId}`, {
      method: 'PATCH',
      body: JSON.stringify(body),
    }),

  /** Delete a milestone */
  deleteMilestone: (milestoneId: number) =>
    fetchApi<{ message: string; id: number }>(`/projects/milestones/${milestoneId}`, {
      method: 'DELETE',
    }),

  /** Assign or unassign a task to a milestone */
  assignTaskToMilestone: (taskId: number, milestoneId: number | null) =>
    fetchApi<{ message: string; task_id: number; milestone_id: number | null }>(
      `/projects/tasks/${taskId}/milestone`,
      {
        method: 'POST',
        params: milestoneId !== null ? { milestone_id: milestoneId } : undefined,
      }
    ),

  // =========================================================================
  // COMMENTS
  // =========================================================================

  /** List comments for an entity */
  getEntityComments: (
    entityType: EntityType,
    entityId: number,
    params?: { limit?: number; offset?: number }
  ) =>
    fetchApi<CommentListResponse>(`/projects/${entityType}/${entityId}/comments`, {
      params: params as Record<string, unknown>,
    }),

  /** Create a comment */
  createComment: (entityType: EntityType, entityId: number, body: CommentCreatePayload) =>
    fetchApi<ProjectComment>(`/projects/${entityType}/${entityId}/comments`, {
      method: 'POST',
      body: JSON.stringify(body),
    }),

  /** Update a comment */
  updateComment: (commentId: number, body: CommentUpdatePayload) =>
    fetchApi<ProjectComment>(`/projects/comments/${commentId}`, {
      method: 'PATCH',
      body: JSON.stringify(body),
    }),

  /** Delete a comment */
  deleteComment: (commentId: number) =>
    fetchApi<{ message: string; id: number }>(`/projects/comments/${commentId}`, {
      method: 'DELETE',
    }),

  // =========================================================================
  // ACTIVITY FEED
  // =========================================================================

  /** List activities for an entity */
  getEntityActivities: (
    entityType: EntityType,
    entityId: number,
    params?: { activity_type?: ProjectActivityType; limit?: number; offset?: number }
  ) =>
    fetchApi<ActivityListResponse>(`/projects/${entityType}/${entityId}/activities`, {
      params: params as Record<string, unknown>,
    }),

  /** Get combined activity timeline for a project */
  getProjectActivityTimeline: (projectId: number, limit?: number) =>
    fetchApi<ActivityTimelineResponse>(`/projects/${projectId}/activity-timeline`, {
      params: limit ? { limit } : undefined,
    }),

  // =========================================================================
  // ATTACHMENTS
  // =========================================================================

  /** List attachments for an entity */
  getEntityAttachments: (entityType: EntityType, entityId: number) =>
    fetchApi<AttachmentListResponse>(`/projects/${entityType}/${entityId}/attachments`),

  /** Upload an attachment */
  uploadAttachment: (
    entityType: EntityType,
    entityId: number,
    file: File,
    options?: { attachment_type?: string; description?: string; is_primary?: boolean }
  ) => {
    const formData = new FormData();
    formData.append('file', file);
    if (options?.attachment_type) formData.append('attachment_type', options.attachment_type);
    if (options?.description) formData.append('description', options.description);
    if (options?.is_primary) formData.append('is_primary', 'true');

    return fetchApi<{ message: string; id: number; file_name: string; file_size: number }>(
      `/projects/${entityType}/${entityId}/attachments`,
      {
        method: 'POST',
        body: formData,
        headers: {}, // Let browser set Content-Type for FormData
      }
    );
  },

  /** Get attachment details */
  getAttachment: (attachmentId: number) =>
    fetchApi<ProjectAttachment>(`/projects/attachments/${attachmentId}`),

  /** Delete an attachment */
  deleteAttachment: (attachmentId: number) =>
    fetchApi<{ message: string; id: number }>(`/projects/attachments/${attachmentId}`, {
      method: 'DELETE',
    }),

  /** Get attachment download URL */
  getAttachmentDownloadUrl: (attachmentId: number) =>
    `/api/projects/attachments/${attachmentId}/download`,

  // =========================================================================
  // APPROVAL WORKFLOWS
  // =========================================================================

  /** Get approval status for a project */
  getProjectApprovalStatus: (projectId: number) =>
    fetchApi<ProjectApprovalStatus>(`/projects/${projectId}/approval-status`),

  /** Submit project for approval */
  submitProjectForApproval: (projectId: number, remarks?: string) =>
    fetchApi<{ message: string; project_id: number; approval_id: number; status: string }>(
      `/projects/${projectId}/submit-approval`,
      {
        method: 'POST',
        body: JSON.stringify({ remarks }),
      }
    ),

  /** Approve a project */
  approveProject: (projectId: number, remarks?: string) =>
    fetchApi<{ message: string; project_id: number; approval_id: number; status: string; current_step: number }>(
      `/projects/${projectId}/approve`,
      {
        method: 'POST',
        body: JSON.stringify({ remarks }),
      }
    ),

  /** Reject a project */
  rejectProject: (projectId: number, reason: string) =>
    fetchApi<{ message: string; project_id: number; approval_id: number; status: string; reason: string }>(
      `/projects/${projectId}/reject`,
      {
        method: 'POST',
        body: JSON.stringify({ reason }),
      }
    ),

  /** Check if current user can approve */
  canApproveProject: (projectId: number) =>
    fetchApi<{ project_id: number; user_id: number; can_approve: boolean }>(
      `/projects/${projectId}/can-approve`
    ),

  // =========================================================================
  // PROJECT TEMPLATES
  // =========================================================================

  /** List all project templates */
  getTemplates: (params?: { active_only?: boolean; is_active?: boolean; project_type?: string }) =>
    fetchApi<ProjectTemplateListResponse>('/projects/templates', {
      params: params as Record<string, unknown>,
    }),

  /** Get a specific template with milestones and tasks */
  getTemplate: (templateId: number) =>
    fetchApi<ProjectTemplate>(`/projects/templates/${templateId}`),

  /** Create a new project template */
  createTemplate: (body: ProjectTemplateCreatePayload) =>
    fetchApi<ProjectTemplate>('/projects/templates', {
      method: 'POST',
      body: JSON.stringify(body),
    }),

  /** Update a project template */
  updateTemplate: (templateId: number, body: ProjectTemplateUpdatePayload) =>
    fetchApi<ProjectTemplate>(`/projects/templates/${templateId}`, {
      method: 'PATCH',
      body: JSON.stringify(body),
    }),

  /** Delete a project template */
  deleteTemplate: (templateId: number) =>
    fetchApi<{ message: string; id: number }>(`/projects/templates/${templateId}`, {
      method: 'DELETE',
    }),

  /** Create a new project from a template */
  createFromTemplate: (templateId: number, body: CreateFromTemplatePayload) =>
    fetchApi<{ message: string; project_id: number; project_name: string; milestones_created: number; tasks_created: number }>(
      `/projects/from-template/${templateId}`,
      {
        method: 'POST',
        body: JSON.stringify(body),
      }
    ),

  // =========================================================================
  // CHANGE HISTORY
  // =========================================================================

  /** Get change history for an entity */
  getEntityHistory: (
    entityType: EntityType,
    entityId: number,
    params?: { limit?: number; offset?: number }
  ) =>
    fetchApi<ChangeHistoryResponse>(`/projects/${entityType}/${entityId}/history`, {
      params: params as Record<string, unknown>,
    }),
};

export default projectsApi;

File: app/api/projects/projects.py
"""
Projects Endpoints
"""

from fastapi import APIRouter, Depends, HTTPException, Query, Body
from sqlalchemy.orm import Session
from typing import Dict, Any, Optional, List
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, Field

from app.database import get_db
from app.auth import Require, Principal, get_current_principal
from app.cache import cached, CACHE_TTL
from app.models import (
    Project,
    ProjectStatus,
    ProjectPriority,
    Task,
    TaskStatus,
)
from app.services.projects import (
    ProjectService,
    ProjectFilters,
)
from app.services.projects.errors import ProjectNotFoundError
from app.services.types import PaginationParams

router = APIRouter()

# =============================================================================
# PROJECTS LIST & DETAIL
# =============================================================================

@router.get("/projects", dependencies=[Depends(Require("explorer:read"))])
async def list_projects(
    status: Optional[str] = None,
    priority: Optional[str] = None,
    customer_account_id: Optional[int] = None,
    project_type: Optional[str] = None,
    department: Optional[str] = None,
    search: Optional[str] = None,
    overdue_only: bool = False,
    start_date: Optional[str] = None,
    end_date: Optional[str] = None,
    limit: int = Query(default=100, le=500),
    offset: int = Query(default=0, ge=0),
    db: Session = Depends(get_db),
    principal: Principal = Depends(get_current_principal),
) -> Dict[str, Any]:
    """List projects with filtering and pagination."""
    # Parse status and priority enums
    status_enum = None
    priority_enum = None

    if status:
        try:
            status_enum = ProjectStatus(status)
        except ValueError:
            raise HTTPException(status_code=400, detail=f"Invalid status: {status}")

    if priority:
        try:
            priority_enum = ProjectPriority(priority)
        except ValueError:
            raise HTTPException(status_code=400, detail=f"Invalid priority: {priority}")

    # Build filters
    filters = ProjectFilters(
        status=status_enum,
        priority=priority_enum,
        customer_account_id=customer_account_id,
        project_type=project_type,
        department=department,
        search=search,
        overdue_only=overdue_only,
        start_date=datetime.fromisoformat(start_date) if start_date else None,
        end_date=datetime.fromisoformat(end_date) if end_date else None,
    )
    pagination = PaginationParams(limit=limit, offset=offset)

    # Use service
    service = ProjectService(db, principal)
    result = service.list_projects(filters, pagination)

    return {
        "total": result.total,
        "limit": result.limit,
        "offset": result.offset,
        "data": [_serialize_project_list(p) for p in result.items],
    }


def _serialize_project_list(p: Project) -> Dict[str, Any]:
    """Serialize a project for list view."""
    return {
        "id": p.id,
        "erpnext_id": p.erpnext_id,
        "project_name": p.project_name,
        "project_type": p.project_type,
        "status": p.status.value if p.status else None,
        "priority": p.priority.value if p.priority else None,
        "department": p.department,
        "customer_account_id": p.customer_account_id,
        "percent_complete": float(p.percent_complete) if p.percent_complete else 0,
        "expected_start_date": p.expected_start_date.isoformat() if p.expected_start_date else None,
        "expected_end_date": p.expected_end_date.isoformat() if p.expected_end_date else None,
        "estimated_costing": float(p.estimated_costing) if p.estimated_costing else 0,
        "total_billed_amount": float(p.total_billed_amount) if p.total_billed_amount else 0,
        "is_overdue": p.is_overdue,
        "task_count": len(p.tasks),
        "created_at": p.created_at.isoformat() if p.created_at else None,
        "write_back_status": getattr(p, "write_back_status", None),
    }


@router.get("/projects/{project_id}", dependencies=[Depends(Require("explorer:read"))])
async def get_project(
    project_id: int,
    db: Session = Depends(get_db),
    principal: Principal = Depends(get_current_principal),
) -> Dict[str, Any]:
    """Get detailed project information with all child tables."""
    service = ProjectService(db, principal)

    try:
        project = service.get_project(project_id)
    except ProjectNotFoundError:
        raise HTTPException(status_code=404, detail="Project not found")

    # Use service methods to get related info
    customer = service.get_customer_info(project)
    manager = service.get_manager_info(project)
    task_stats = service.get_project_task_stats(project_id)

    return _serialize_project_detail(project, customer, manager, task_stats)


def _serialize_project_detail(
    project: Project,
    customer: Optional[Dict[str, Any]],
    manager: Optional[Dict[str, Any]],
    task_stats,
) -> Dict[str, Any]:
    """Serialize a project for detail view."""
    # Build users list (team members)
    users = [
        {
            "id": u.id,
            "user": u.user,
            "full_name": u.full_name,
            "email": u.email,
            "project_status": u.project_status,
        }
        for u in sorted(project.users, key=lambda x: x.idx or 0)
    ]

    # Build tasks list
    tasks = [
        {
            "id": t.id,
            "erpnext_id": t.erpnext_id,
            "subject": t.subject,
            "status": t.status.value if t.status else None,
            "priority": t.priority.value if t.priority else None,
            "assigned_to": t.assigned_to,
            "progress": float(t.progress) if t.progress else 0,
            "exp_start_date": t.exp_start_date.isoformat() if t.exp_start_date else None,
            "exp_end_date": t.exp_end_date.isoformat() if t.exp_end_date else None,
            "is_overdue": t.is_overdue,
            "dependency_count": len(t.depends_on),
        }
        for t in project.tasks
    ]

    # Build expenses list
    expenses = [
        {
            "id": e.id,
            "erpnext_id": e.erpnext_id,
            "expense_type": e.expense_type,
            "description": e.description,
            "total_claimed_amount": float(e.total_claimed_amount) if e.total_claimed_amount else 0,
            "total_sanctioned_amount": float(e.total_sanctioned_amount) if e.total_sanctioned_amount else 0,
            "status": e.status.value if e.status else None,
            "expense_date": e.expense_date.isoformat() if e.expense_date else None,
            "employee_name": e.employee_name,
        }
        for e in project.expenses
    ]

    return {
        "id": project.id,
        "erpnext_id": project.erpnext_id,
        "project_name": project.project_name,
        "project_type": project.project_type,
        "status": project.status.value if project.status else None,
        "priority": project.priority.value if project.priority else None,
        "department": project.department,
        "company": project.company,
        "cost_center": project.cost_center,
        "percent_complete": float(project.percent_complete) if project.percent_complete else 0,
        "gross_margin": float(project.gross_margin) if project.gross_margin else 0,
        "total_sales_amount": float(project.total_sales_amount) if project.total_sales_amount else 0,
        "progress": {
            "percent_complete": float(project.percent_complete) if project.percent_complete else 0,
            "percent_complete_method": project.percent_complete_method,
            "is_active": project.is_active,
        },
        "dates": {
            "expected_start_date": project.expected_start_date.isoformat() if project.expected_start_date else None,
            "expected_end_date": project.expected_end_date.isoformat() if project.expected_end_date else None,
            "actual_start_date": project.actual_start_date.isoformat() if project.actual_start_date else None,
            "actual_end_date": project.actual_end_date.isoformat() if project.actual_end_date else None,
            "created_at": project.created_at.isoformat() if project.created_at else None,
            "updated_at": project.updated_at.isoformat() if project.updated_at else None,
        },
        "financials": {
            "estimated_costing": float(project.estimated_costing) if project.estimated_costing else 0,
            "total_costing_amount": float(project.total_costing_amount) if project.total_costing_amount else 0,
            "total_expense_claim": float(project.total_expense_claim) if project.total_expense_claim else 0,
            "total_purchase_cost": float(project.total_purchase_cost) if project.total_purchase_cost else 0,
            "total_sales_amount": float(project.total_sales_amount) if project.total_sales_amount else 0,
            "total_billable_amount": float(project.total_billable_amount) if project.total_billable_amount else 0,
            "total_billed_amount": float(project.total_billed_amount) if project.total_billed_amount else 0,
            "gross_margin": float(project.gross_margin) if project.gross_margin else 0,
            "profit_margin_percent": float(project.profit_margin_percent),
        },
        "time_tracking": {
            "actual_time": float(project.actual_time) if project.actual_time else 0,
            "total_consumed_material_cost": float(project.total_consumed_material_cost) if project.total_consumed_material_cost else 0,
        },
        "notes": project.notes,
        "is_overdue": project.is_overdue,
        "customer": customer,
        "project_manager": manager,
        "users": users,
        "tasks": tasks,
        "task_stats": {
            "total": task_stats.total,
            "completed": task_stats.completed,
            "open": task_stats.open,
            "overdue": task_stats.overdue,
        },
        "expenses": expenses,
        "write_back_status": getattr(project, "write_back_status", None),
    }

